ngraph/test/type_prop/matmul.cpp

```diff
@@ -108,3 +119,408 @@ TEST(type_prop, matmul_4D_transpose_b)
+    auto B =
+        make_shared<op::Parameter>(element::f32, PartialShape{Dimension(7, 8), Dimension(15, 20)});
+
+    auto matmul = make_shared<op::MatMul>(A, B, false, false);
+
+    ASSERT_EQ(matmul->get_element_type(), element::f32);
+    ASSERT_EQ(matmul->get_output_partial_shape(0),
+              (PartialShape{Dimension(2, 5), Dimension(15, 20)}));
+}
+
+TEST(type_prop, matmul_batch_dynamic_bounds)
+{
+    // Input A and input B dim bounds => output dim bound
+    // Dimension 1 can be expanded to any bigger
+
+    Dimension dynamic = Dimension::dynamic();
+
+    auto A_shape = PartialShape{dynamic,          // 0
+                                Dimension(1, 5),  // 1
+                                Dimension(2, 10), // 2
+                                Dimension(5, 7),  // 3
+                                Dimension(4, 7),  // 4
+                                Dimension(5, 10), // 5
+                                Dimension(1, 4),  // 6
+                                Dimension(0, 1),  // 7
+                                Dimension(0, 3),  // 8
+                                1,                // 9
+                                1,                // 10
+                                5,                // 11
+                                6};               // 12
+
+    auto B_shape = PartialShape{dynamic,           // 0
+                                Dimension(10, 20), // 1
+                                Dimension(10, 20), // 2
+                                Dimension(4, 10),  // 3
+                                Dimension(5, 10),  // 4
+                                Dimension(4, 7),   // 5
+                                dynamic,           // 6
+                                Dimension(0, 1),   // 7
+                                Dimension(2, 5),   // 8
+                                dynamic,           // 9
+                                3,                 // 10
+                                6,                 // 11
+                                4};                // 12
+
+    auto expected_output_shape = PartialShape{dynamic,           // 0
+                                              Dimension(10, 20), // 1
+                                              10,                // 2
+                                              Dimension(5, 7),   // 3
+                                              Dimension(5, 7),   // 4
+                                              Dimension(5, 7),   // 5
+                                              Dimension(1, -1),  // 6
+                                              Dimension(0, 1),   // 7
+                                              Dimension(2, 5),   // 8
+                                              Dimension(1, -1),  // 9
+                                              3,                 // 10
+                                              5,                 // 11
+                                              4};                // 12
+
```

Comment:
@slyalin Please confirm that dynamic dimension range tests match common expectations.

Reply:
Particularly `[0, 1] & [0, 1]` case, we can infer as output dim with `[0, 1]` range
or consider as static dim with single value `1` because `0` would be wrong batch dimension. 
Same way `dynamic & dynamic` which is `[0, max] & [0, max]` can be just `dynamic` or `[1, max]`.

(Init Dimension with  `-1` means Interval min/max value limit)

Reply:
0 is a valid dimension value which means that the corresponding tensor is empty, but still have some dimensions. We don't have any exceptions for 0 dimensions in the specification. So we shouldn't ignore 0 by treating it as impossible.